File: 3rd Semester/Advanced Programming Methods(Done)/Laboratory/src (BUX)/Model/Utils/MyStack.java
package Model.Utils;
import Exceptions.MyStackException;
import Model.Expressions.Exp;
import Model.Statements.IStmt;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Stack;


public class MyStack<T> implements MyIStack<T> {

    private Stack<T> stk;
    public MyStack()
    {
        stk=new Stack<T>();
    }


    public void push(T el){
        stk.push(el);
    }

    public String toString(){
        List<T> stack= new ArrayList<>(stk);
        Collections.reverse(stack);
        String msg= "";
        for(T el:stack){
            msg+=el+";";
        }return msg;
    }
    public T pop(){
        return stk.pop();
    }
    public void isEmpty() throws MyStackException{
        if (stk.isEmpty())
            throw new MyStackException("Empty stack!");
    }
    public Stack<T> getStk(){
        return stk;
    }
    public void check(Exp exp, IStmt stmt){
        if (stk.pop()==exp)
            stk.push((T) stmt);

    }
}

File: 3rd Semester/Advanced Programming Methods/Laboratory/Toy Interpreter C#/sharplab/Model/Utils/MyStack.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace sharplab.Model.Utils
{
    class MyStack<T> : IMyStack<T>
    {
        private Stack<T> stk;
        public MyStack()
        {
            stk = new Stack<T>();
        }
        public void Push(T el)
        {
            stk.Push(el);
        }

        public override string ToString()
        {
            String msg = "";
            foreach (T el in stk)
            {
                msg += el + ";";
            }
            return msg;
        }
        public T Pop()
        {
            return stk.Pop();
        }

        public Stack<T> GetStk()
        {
            return stk;
        }

        public void IsEmpty()
        {
            if (stk.Count == 0)
                throw new Exception("Stack is empty!");
        }
    }
}
